extensions/amp-analytics/0.1/variables.js

```diff
@@ -0,0 +1,298 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {cryptoFor} from '../../../src/crypto';
+import {dev, user} from '../../../src/log';
+import {fromClass} from '../../../src/service';
+import {isArray, isFiniteNumber, map} from '../../../src/types';
+
+/** @const {string} */
+const TAG = 'Analytics.Variables';
+
+/** @const {RegExp} */
+const VARIABLE_ARGS_REGEXP = /^(?:([^ ]*)(\([^)]*\))|.+)$/;
+
+/**
+ * @struct
+ * @const
+ */
+class FilterDef {
+  /**
+   * @param {function(...?):(?|!Promise<?>)} filter
+   * @param {boolean=} opt_allowNull
+   */
+  constructor(filter, opt_allowNull) {
+    /** @type {!function(...?):(?|!Promise<?>)} */
+    this.filter = filter;
+
+    /** @type{boolean} */
+    this.allowNull = !!opt_allowNull;
+  }
+}
+
+
+/**
+ * The structure that contains all details needed to expand a template
+ * @struct
+ * @const
+ * @package For type.
+ */
+export class ExpansionOptions {
+  /**
+   * @param {!Object<string, *>} vars
+   * @param {number=} opt_iterations
+   * @param {boolean=} opt_noEncode
+   */
+  constructor(vars, opt_iterations, opt_noEncode) {
+    /** @const {!Object<string, string|Array<string>>} */
+    this.vars = vars;
+    /** @const {!number} */
+    this.iterations = opt_iterations === undefined ? 2 : opt_iterations;
+    /** @const {boolean} */
+    this.noEncode = !!opt_noEncode;
+  }
+}
+
+
+
+/**
+ * @param {!string} str
+ * @param {Number} s
```

Comment:
Commented before: since the real args are all strings and we are doing type conversion below - let's just go ahead and type them as strings.